node-problem-detector: report disk queue length in Prometheus format

xueweiz opened the pull request:
This change is to address below items from #284 :

- [x] Refactor problem daemon registration and initialization, so that they are more modular, and rely on less per-daemon initialization code duplication.
- [x] Refactor events/condition reporting into k8s_exporter.
- [x] Add a SystemStatsMonitor problem daemon, to monitor some system metrics; and add an optional Prometheus metrics exporter.

Since this PR will need OpenCensus and gopsutil library, I submitted PR #289 for it.

Deprecated (but not removed yet) flags in this PR:

- --custom-plugin-monitors
- --system-log-monitors

New flags introduced in this PR:

- --config.custom-plugin-monitor
- --config.system-log-monitor
- --config.system-stats-monitor
- --enable-k8s-exporter
- --prometheus-address
- --prometheus-port

Test steps:

1. Verifying existing command works:
> make ./bin/node-problem-detector && ./bin/node-problem-detector --system-log-monitors=config/kernel-monitor.json --apiserver-override=http://127.0.0.1:8080?inClusterConfig=false --logtostderr

2. Testing new flags for existing behavior:
> make ./bin/node-problem-detector && ./bin/node-problem-detector --config.system-log-monitor=config/kernel-monitor.json --apiserver-override=http://127.0.0.1:8080?inClusterConfig=false --logtostderr

3. Test NPD for disk monitor with no k8s_exporter
> make ./bin/node-problem-detector && ./bin/node-problem-detector --config.system-log-monitor=config/kernel-monitor.json --apiserver-override=http://127.0.0.1:8080?inClusterConfig=false --logtostderr --config.system-stats-monitor=config/system-stats-monitor.json --enable-k8s-exporter=false

> curl http://localhost:20257/metrics

Follow up:
I will later update test infra to use these new flags.

xueweiz commented:
Just signed the CLA agreement.

xueweiz commented:
/assign Random-Liu@

xueweiz commented:
Sorry, wrong assigning format.
/assign @Random-Liu 

Hi Lantao, could you help review this PR? Thanks!
The first 5 changes are all Godep management, only the 6th one is code change.

xueweiz commented:
The change can be tested via:
kubectl proxy --port=8080 > /dev/null &
make ./bin/node-problem-detector && ./bin/node-problem-detector --system-log-monitors=config/kernel-monitor.json --apiserver-override=http://127.0.0.1:8080?inClusterConfig=false --logtostderr
curl http://localhost:20257/metrics

I wanted to add some unit tests, but I don't think they are very useful for this PR (sure I could mock the lsblk / OpenCensus / gopsutil dependencies, but those are ~100% interaction testing and ~0% behavior testing). I think integration tests will be far more useful.

Please let me know if you feel some part should be unit-tested. I could try to re-organize the code a bit to test them in mocking. Thanks! :smiley:

Random-Liu commented:
/ok-to-test

andyxning commented:
@xueweiz Could you please give us(maybe @Random-Liu has known about this) some background about adding this feature? Maybe a tiny doc is ok. This makes features trackable just like when we add custom plugin support.

andyxning commented:
Seems flakes. 

/test pull-npd-test

xueweiz commented:
@andyxning Hi Ning, thanks for taking a look of the PR!
Yes I plan to send out a formal design doc similar to the custom plugin one. I'd still need a bit time to polish the doc, sorry for the delay.

Do you mind if I first send out a short version of the design doc, and populate it with more details later on? Thanks!

xueweiz commented:
> Seems flakes.
> 
> /test pull-npd-test

Hum interesting. In both test runs, the failure is at TestGoroutineLeak at log_monitor_test.go. The PR did not touch that code path (maybe the vendor/ changes caused that?)
What's even weirder, is that the goroutine is not leaking, it is disappearing...At the beginning of the test we have 5 goroutines, at the end of test we only have 4.

Either way, I'll add some stacktrace printing on this testcase, to help current/future debugging.

I just force pushed the PR with changes to print of stacktraces in this testcase.

/test pull-npd-test

xueweiz commented:
Interesting. From the [last test result](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/node-problem-detector/275/pull-npd-test/1128378682335825920)
It claims that originally there were 5 goroutines, although `pprof.Lookup("goroutine")` only showed 4 goroutines.

Seems one goroutine finished right in the gap (after `runtime.NumGoroutine()` and before `pprof.Lookup("goroutine")`). So let's stop using two calls to retrieve goroutines, but use only one.

Let's see how it goes :)

/test pull-npd-test

andyxning commented:
@xueweiz 
> Do you mind if I first send out a short version of the design doc, and populate it with more details later on? Thanks!

Yes. Short version is helpful for us to review the motivation.

xueweiz commented:
@andyxning 
Hi Ning, I just finished writing up the [design doc](https://docs.google.com/document/d/1SeaUz6kBavI283Dq8GBpoEUDrHA2a795xtw0OvjM568/edit?usp=sharing): https://docs.google.com/document/d/1SeaUz6kBavI283Dq8GBpoEUDrHA2a795xtw0OvjM568/edit?usp=sharing

Sorry for the delay. It'd be great if you can take a look and provide some feedback :) Thanks a lot!

andyxning commented:
Good. Will take a look at it asap.

xueweiz commented:
@andyxning  @Random-Liu 

Hi Ning, Lantao,

I just updated the code to separate works into multiple commits, so that it's easier to review.

This pull request implements parts of the [proposal](https://docs.google.com/document/d/1SeaUz6kBavI283Dq8GBpoEUDrHA2a795xtw0OvjM568/edit#heading=h.xw2rl6ej7zhr).

I will wait until this PR has some initial review before starting the other half (e.g. converting existing problems into metrics) until.

xueweiz commented:
This is strange. I see the TestGoroutineLeak failure at [here](https://travis-ci.org/kubernetes/node-problem-detector/jobs/535073809).

But the strange thing is that the printed detailed stack matches our expectation...Somehow `original_prof.Count()` thinks there were 5 goroutine, while the printing clearly shows that there were only 4:

        original stacktraces: goroutine 5 [running]:
        runtime/pprof.writeGoroutineStacks(0x106eec0, 0xc0004745a0, 0x1000000000000, 0xd0)
        
        goroutine 1 [chan receive]:
        testing.(*T).Run(0xc000270b00, 0xf47e95, 0x11, 0xf8fe30, 0x1)
        
        goroutine 19 [chan receive]:
        k8s.io/node-problem-detector/vendor/github.com/golang/glog.(*loggingT).flushDaemon(0x18b2ea0)

        goroutine 23 [select]:
        k8s.io/node-problem-detector/vendor/go.opencensus.io/stats/view.(*worker).start(0xc000150780)

CodeJuan commented on pkg/exporter/prometheus_exporter.go:
```diff
@@ -0,0 +1,60 @@
+}
+
+// ExportProblems does nothing.
+// Prometheus exporter ignores all problems, and only deal with metrics.
```
I'm not sure ignores all problems is a good idea, I think problem is a kind of metrics.
There is a case: The remedy system gets problem from prometheus or other monitoring system, then start remedy job, such as restart some processes, reboot node. 
So if NPD could export the problem as metric would be much helpful.

xueweiz replied:
Hi Juan,

Thanks for the review! And I totally agree that NPD should export all current problems as metrics.
The detailed design of how I plan to achieve that is at the "[Extending Today’s Problem Daemon to Report Metrics](https://docs.google.com/document/d/1SeaUz6kBavI283Dq8GBpoEUDrHA2a795xtw0OvjM568/edit#heading=h.eym0m8duezae)" section of the [proposal](https://docs.google.com/document/d/1SeaUz6kBavI283Dq8GBpoEUDrHA2a795xtw0OvjM568/edit).

The basic idea is, problem daemons (today we only have two, System Log Monitor and Custom Plugin Monitor) will convert problems to metrics (counter, or gauge). For example:
`problem_counter{"name": "DockerRestart"}  17`

The current [config files](https://github.com/kubernetes/node-problem-detector/tree/v0.6.3/config) won't need any modification to support this. Default metrics conversion will be provided at problem daemon level. Of course, if fancier conversion is needed (e.g. renaming the metric name/label), we could allow those to be configurable as well.

Since problem daemons will convert problems into metrics and report metrics directly, Prometheus exporter can just happy take the _metrics_, and ignore all _problems_ :)

The line `// Prometheus exporter ignores all problems, and only deal with metrics` actually means that Prometheus exporter do not report _problems_, but only report _metrics_. The notion of _problems_ and _metrics_ is newly introduced in the [proposal](https://docs.google.com/document/d/1SeaUz6kBavI283Dq8GBpoEUDrHA2a795xtw0OvjM568/edit). Please see the "[Detailed NPD architecture](https://docs.google.com/document/d/1SeaUz6kBavI283Dq8GBpoEUDrHA2a795xtw0OvjM568/edit#heading=h.jhsmauthxd88)" and "[Core Data Representation Model](https://docs.google.com/document/d/1SeaUz6kBavI283Dq8GBpoEUDrHA2a795xtw0OvjM568/edit#heading=h.tlefaa9eq3qb)" sections.

CodeJuan replied:
Thank you for prompt reply. It make sense to me.

xueweiz replied:
Thanks Juan :)

Random-Liu replied:
Then please change the comment. :)

xueweiz replied:
Hum...I feel the comment was correct...
Changed it to `// Prometheus exporter only exports metrics.`
Do you think this would reduce the confusion? Thanks!

Random-Liu commented on cmd/node_problem_detector.go:
```diff
@@ -46,23 +47,27 @@ func main() {
+	for _, err := range errors {
+		glog.Errorf("Problem daemon initialization failed: %v", err)
+	}
+	if len(errors) != 0 && npdo.PanicOnProblemDaemonFailure {
```
Why not just always panic if a problem daemon fails? Any concern about that?

xueweiz replied:
Sounds fair :)
Changed the behavior to just panic.

Random-Liu commented on cmd/node_problem_detector.go:
```diff
@@ -46,23 +47,27 @@ func main() {
+	// Legacy per-monitor configuration.
+	// This section is kept so that existing command line option still works on NPD.
+	// TODO(xueweiz): when NPD's refactoring is finished, remove legacy configuration.
+	for _, config := range npdo.SystemLogMonitorConfigPaths {
```
Why not just change original flags to write to new monitor config?
So that we can preserve original flags, but all use the new monitor config internally.

xueweiz replied:
Actually I can't. The flag that's defined at last will always overwrite the previously defined flag.

For example:

```
        for _, problemDaemonName := range problemdaemon.GetProblemDaemonNames() {
		npdo.MonitorsConfigs[problemDaemonName] = &[]string{}
		fs.StringSliceVar(
			npdo.MonitorsConfigs[problemDaemonName],
			"config."+string(problemDaemonName),
			[]string{},
			fmt.Sprintf("Comma separated configurations for %v monitor. %v",
				problemDaemonName,
				problemdaemon.GetProblemDaemonHandler(problemDaemonName).CmdOptionDescription))
	}
	fs.StringSliceVar(npdo.MonitorsConfigs["system-log-monitor"], "system-log-monitors",
		[]string{}, "List of paths to system log monitor config files, comma separated. This option will be replaced by --config.system-log-monitor.")
	fs.StringSliceVar(npdo.MonitorsConfigs["system-log-monitor"], "custom-plugin-monitors",
		[]string{}, "List of paths to custom plugin monitor config files, comma separated. This option will be replaced by --config.custom-plugin-monitor.")
```

In this code, "--config.system-log-monitor" argument will always be ignored. Because whatever the value it has, "--system-log-monitors" will always overwrite it.

I think it's OK if we leave this code for now. I plan to first change all the test to use the new argument, then remove this code. It's shouldn't be too bad.

wangzhen127 replied:
> In this code, "--config.system-log-monitor" argument will always be ignored. Because whatever the value it has, "--system-log-monitors" will always overwrite it.

I don't understand this part. We can update the code to change the behavior.

My preferred behavior is: panic if both `--config.system-log-monitor` and `--system-log-monitors` are specified. We should only allow one is used. And then here, we can achieve what @Random-Liu described: change original flags to write to the new monitor config.

xueweiz replied:
Good point, thanks! I like the panic behavior.
Change the code so that when both old and new flags are used, NPD will panic. And added unit test for it.

wangzhen127 replied:
I think @Random-Liu's original comment on "Why not just change original flags to write to new monitor config?" is still valid.

if `--system-log-monitor` is specified, set it to `--config.system-log-monitor`. Then here, we can just have one line:
```
problemDaemons := problemdaemon.NewProblemDaemons(npdo.MonitorConfigPaths)
```
The legacy code of creating problem daemons from old flags can be removed.

xueweiz replied:
Ah good point. Followed your setConfigsOrDie() idea. Done.

Random-Liu commented on cmd/node_problem_detector.go:
```diff
@@ -46,23 +47,27 @@ func main() {
+	for _, config := range npdo.SystemLogMonitorConfigPaths {
+		problemDaemons = append(problemDaemons, systemlogmonitor.NewLogMonitorOrDie(config))
+	}
 	for _, config := range npdo.CustomPluginMonitorConfigPaths {
```
ditto.

xueweiz replied:
Will do if we decide to merge the argument holders.

Random-Liu commented on pkg/problemdaemon/custom_plugin_monitor.go:
```diff
@@ -0,0 +1,30 @@
+)
+
+func init() {
+	register(
```
Make this public, and move into its own corresponding package?
Basically keep all custom plugin monitor logic in its own package.

xueweiz replied:
Sounds good. Thanks!

Random-Liu commented on pkg/custompluginmonitor/custom_plugin_monitor.go:
```diff
@@ -41,35 +41,43 @@ type customPluginMonitor struct {
 
 // NewCustomPluginMonitorOrDie create a new customPluginMonitor, panic if error occurs.
 func NewCustomPluginMonitorOrDie(configPath string) types.Monitor {
```
We can probably remove this once you redirect the flags above.

xueweiz replied:
Done. Thanks!

Random-Liu commented on pkg/problemdaemon/system_log_monitor.go:
```diff
@@ -0,0 +1,30 @@
+)
+
+func init() {
+	register(
```
ditto.

xueweiz replied:
Done.

Random-Liu commented on pkg/systemlogmonitor/log_monitor.go:
```diff
@@ -44,29 +44,37 @@ type logMonitor struct {
 
 // NewLogMonitorOrDie create a new LogMonitor, panic if error occurs.
 func NewLogMonitorOrDie(configPath string) types.Monitor {
```
ditto.

xueweiz replied:
Done.

Random-Liu commented on pkg/problemdaemon/disk.go:
```diff
@@ -0,0 +1,164 @@
+/*
```
I initially thought that you'll have another level of plugin mechanism for compile time plug-in metrics, but it seems that you decide to reuse the problem daemon plugin mechanism.

If this is the case, can:
1) Move the logic into its own package (assuming Register is public);
2) Consider whether we just want this to be a "disk monitor" or "system stats monitor". Let's say we need some memory stats in the future, do we want to add another problem daemon?

xueweiz replied:
> Do you plan to provide a compile time mechanism to exclude problem daemons that are not needed for some users?

Yes. I have two ideas to achieve that, but not decided yet. I plan to submit a following PR to address the "Implement option for disabling problem daemons at compile time" action item at #284 

I don't think it's very urgent yet. So planning to do it in a follow up PR.

> Consider whether we just want this to be a "disk monitor" or "system stats monitor". 

> A monolithic "system stats monitor" cannot satisfy such requirement.

I plan to introduce a config file to select the collected metrics anyway. So we can achieve the granularity through the config file.

I'm calling the problem daemon "system stats monitor" for now. We can allow it to evolve over time.

Random-Liu commented:
Are you going to send another PR to add the problem_count support? Or do you plan to do that in this PR?

jiayingz commented:
/cc @jiayingz @yguo0905

wangzhen127 commented on cmd/options/options.go:
```diff
@@ -63,6 +69,7 @@ func (npdo *NodeProblemDetectorOptions) AddFlags(fs *pflag.FlagSet) {
 		[]string{}, "List of paths to system log monitor config files, comma separated.")
 	fs.StringSliceVar(&npdo.CustomPluginMonitorConfigPaths, "custom-plugin-monitors",
 		[]string{}, "List of paths to custom plugin monitor config files, comma separated.")
+	fs.BoolVar(&npdo.EnableK8sExporter, "enable-k8s-exporter", true, "Enables reporting to Kubernetes.")
```
What happens when `enable-k8s-exporter` is false? I don't see `enable-k8s-exporter` used anywhere.

xueweiz replied:
In `NewK8SExporterOrDie()`, it will return `nil` if this flag is false.

xueweiz commented on pkg/problemdaemon/problem_daemon.go:
```diff
@@ -0,0 +1,62 @@
+)
+
+// register registers a problem daemon factory method, which will be used to create the problem daemon.
+func register(problemDaemonType types.ProblemDaemonType, handler types.ProblemDaemonHandler) {
```
Thanks for the tips! Changed.

xueweiz commented on pkg/types/types.go:
```diff
@@ -113,3 +113,20 @@ type Exporter interface {
+// ProblemDaemonHandler represents the initialization handler for a type problem daemon.
+type ProblemDaemonHandler struct {
+	// The method used to initialize a problem daemon instance.
+	Factory func(string) (Monitor, error)
```
Thanks!

xueweiz commented on pkg/types/types.go:
```diff
@@ -113,3 +113,20 @@ type Exporter interface {
+	// The method used to initialize a problem daemon instance.
+	Factory func(string) (Monitor, error)
+	// The command line description about how to configure the problem daemon.
+	CmdOptionDescription string
```
Thanks!

xueweiz commented on cmd/options/options.go:
```diff
@@ -36,8 +36,6 @@ type NodeProblemDetectorOptions struct {
 	// CustomPluginMonitorConfigPaths specifies the list of paths to custom plugin monitor configuration
 	// files.
 	CustomPluginMonitorConfigPaths []string
```
Yes. I plan to do that in a future, in #284 action item:

> Implement option for disabling exporters at compile time.

The main challenge is how would the command line option looks like.
I don't think it's very urgent right now.

xueweiz commented on pkg/problemdaemon/disk.go:
```diff
@@ -0,0 +1,164 @@
+
+// NewDiskCollector creates a new disk metrics collector.
+func NewDiskCollector(config string) (types.Monitor, error) {
+	if config != "true" {
```
Sounds fair. I want it to point to a config file anyway. So let's do that :)
I'll follow up with the change soon.

xueweiz replied:
Done.

xueweiz commented on cmd/node_problem_detector.go:
```diff
@@ -72,6 +72,9 @@ func main() {
 
 	// Initialize exporters.
 	exporters := []types.Exporter{}
+	if pe := exporter.NewPrometheusExporterOrDie(npdo); pe != nil {
```
When `npdo.PrometheusServerPort` <= 0, `pe` will be `nil`. That implements the option to control it.

I plan to allow compile disabling as well. But let's do that in another PR. As it requires a bit design as well.

xueweiz commented:
> Are you going to send another PR to add the problem_count support? Or do you plan to do that in this PR?

Yes. I plan to do it in another PR.

xueweiz commented:
The automated test failed because it couldn't find OpenCensus library.
I added the library to Godep at #286 , so that this PR will be cleaner.
We could merge #286 first, then the automated test on this PR would work.

wangzhen127 commented on cmd/options/options.go:
```diff
@@ -63,6 +69,7 @@ func (npdo *NodeProblemDetectorOptions) AddFlags(fs *pflag.FlagSet) {
 	fs.StringSliceVar(&npdo.CustomPluginMonitorConfigPaths, "custom-plugin-monitors",
 		[]string{}, "List of paths to custom plugin monitor config files, comma separated.")
+	fs.BoolVar(&npdo.EnableK8sExporter, "enable-k8s-exporter", true, "Enables reporting to Kubernetes.")
 	fs.StringVar(&npdo.ApiServerOverride, "apiserver-override",
```
If k8s exporter is not enabled, this is not necessary. `ValidOrDie()` needs update.

Also, update the description to state that this is ignored if `--enabled-k8s-exporter` is false.

xueweiz replied:
Good catch! Done.

wangzhen127 commented on cmd/options/options.go:
```diff
@@ -63,6 +69,7 @@ func (npdo *NodeProblemDetectorOptions) AddFlags(fs *pflag.FlagSet) {
 		[]string{}, "List of paths to system log monitor config files, comma separated.")
 	fs.StringSliceVar(&npdo.CustomPluginMonitorConfigPaths, "custom-plugin-monitors",
 		[]string{}, "List of paths to custom plugin monitor config files, comma separated.")
+	fs.BoolVar(&npdo.EnableK8sExporter, "enable-k8s-exporter", true, "Enables reporting to Kubernetes.")
```
Description: Enables exporting to Kubernetes API server.

xueweiz replied:
Done.

wangzhen127 commented on pkg/k8sexporter/k8s_exporter.go:
```diff
@@ -0,0 +1,92 @@
+		return nil
+	}
+
+	c := problemclient.NewClientOrDie(npdo)
```
problemclient package should be renamed. That is just for k8s now. In terms of directory structure. Create `/pkg/exporters`. Put k8s and prometheus exporters into that.

xueweiz replied:
Good point. Done.

wangzhen127 commented on pkg/prometheusexporter/prometheus_exporter.go:
```diff
@@ -0,0 +1,60 @@
+/*
```
Create `/pkg/exporters/` and put both k8s and prometheus exporters into it.

xueweiz replied:
Done.

wangzhen127 commented on cmd/options/options.go:
```diff
@@ -90,6 +109,12 @@ func (npdo *NodeProblemDetectorOptions) ValidOrDie() {
 	if len(npdo.SystemLogMonitorConfigPaths) == 0 && len(npdo.CustomPluginMonitorConfigPaths) == 0 {
 		panic(fmt.Sprintf("Either --system-log-monitors or --custom-plugin-monitors is required"))
 	}
+
```
Ah, I see you removed this part in a later commit. I think we should keep this, but update it to check there should be at least one problem daemon configuration. `SystemLogMonitorConfigPaths` can be considered the same as `--config.system-log-monitor`.

xueweiz replied:
Yes, we still have the "at least one problem daemon configuration" check. It is added at "cmd/node_problem_detector.go" at commit "Add existing monitors into the problem daemon registration hook":
```
if len(problemDaemons) == 0 {
		glog.Fatalf("No problem daemon is successfully setup")
}
```

wangzhen127 replied:
It makes more sense to do the flag checks before creating problem daemons. If we set the value of `--system-log-monitor` to `--config.system-log-monitor`, we can keep this check here.

So maybe you can add a function called `setConfigsOrDie()`. In that function:
1. make sure old and new flags do not appear at the same time.
2. If old flag is used, set the value to the new flag and remove the value in old flag.
3. Panic if no config is specified.
3. Later when creating problem daemons, you can just create the problem daemons using the new flag.

xueweiz replied:
Done.

wangzhen127 commented on pkg/problemdetector/problem_detector.go:
```diff
@@ -47,17 +45,16 @@ func NewProblemDetector(monitors map[string]types.Monitor, exporters []types.Exp
 		if err != nil {
 			// Do not return error and keep on trying the following config files.
-			glog.Errorf("Failed to start log monitor %q: %v", cfg, err)
+			glog.Errorf("Failed to start log monitor %v: %v", m, err)
```
s/log monitor/monitor

xueweiz replied:
Good catch! Done.

wangzhen127 commented on pkg/problemdetector/problem_detector.go:
```diff
@@ -47,17 +45,16 @@ func NewProblemDetector(monitors map[string]types.Monitor, exporters []types.Exp
 		}
-		chans = append(chans, ch)
-	}
-	if len(chans) == 0 {
```
Why is this removed?

xueweiz replied:
This logic is moved to "cmd/node_problem_detector.go" in this commit:
```
if len(problemDaemons) == 0 {
	glog.Fatalf("No problem daemon is successfully setup")
}
```

xueweiz replied:
Good point. Done.

wangzhen127 commented on cmd/node_problem_detector.go:
```diff
@@ -69,6 +70,9 @@ func main() {
 		exporters = append(exporters, ke)
 	}
+	if pe := prometheusexporter.NewExporterOrDie(npdo); pe != nil {
+		exporters = append(exporters, pe)
```
Log somewhere when an exporter is successfully started. This applies to all exporters.

xueweiz replied:
Done.

wangzhen127 commented on cmd/options/options.go:
```diff
@@ -88,6 +94,11 @@ func (npdo *NodeProblemDetectorOptions) AddFlags(fs *pflag.FlagSet) {
 		"127.0.0.1", "The address to bind the node problem detector server.")
 
+	fs.IntVar(&npdo.PrometheusServerPort, "prometheus-port",
+		20257, "The port to bind the Prometheus scrape endpoint. Use 0 to disable.")
```
Update the description to explicitly state prometheus exporter is enabled by default, and use 0 to disable.

xueweiz replied:
Done.

wangzhen127 commented on config/system-stats-monitor.json:
```diff
@@ -0,0 +1,19 @@
+		"includeAllAttachedBlk": true,
+		"lsblkTimeout": "5s"
+	},
+	"invokeInterval": "60s"
```
Do you expect all collectors in system stats monitor should use the same invoke interval?

xueweiz replied:
For now, yes, for simplicity.
If in future we want to allow different "sub system stats monitors" to use different interval, we could easily add another config option to allow it.

wangzhen127 commented on pkg/systemstatsmonitor/system_stats_monitor.go:
```diff
@@ -0,0 +1,110 @@
+	// Apply configurations.
+	f, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		glog.Fatalf("Failed to read configuration file %q: %v", configPath, err)
```
When printing fatal log messages, also indicate this is for system stats monitor. This applies to all the fatal messages in this file.

xueweiz replied:
I think it's probably cleaner to not print this. Because [glog.Fatalf](https://godoc.org/github.com/golang/glog#Fatalf) already prints out the stack trace (which tells people that system stats monitor is having trouble), and we also printed out the config file path. It should be pretty intuitive to figure out which problem daemon is failing.

xueweiz commented on cmd/options/options.go:
```diff
@@ -53,14 +50,25 @@ type NodeProblemDetectorOptions struct {
+	// files.
+	CustomPluginMonitorConfigPaths []string
+	// MonitorsConfigs specifies the list of paths to configuration files for each monitor.
+	MonitorsConfigs types.ProblemDaemonConfigMap
```
Done.

xueweiz commented on pkg/types/types.go:
```diff
@@ -113,3 +113,20 @@ type Exporter interface {
+
+// ProblemDaemonConfigMap represents configurations on all types of problem daemons:
+// 1) Each key represents a type of problem daemon.
+// 2) Each value represents the configurations to that type of problem daemon.
```
Done.

xueweiz commented on pkg/types/types.go:
```diff
@@ -113,3 +113,20 @@ type Exporter interface {
+// ProblemDaemonConfigMap represents configurations on all types of problem daemons:
+// 1) Each key represents a type of problem daemon.
+// 2) Each value represents the configurations to that type of problem daemon.
+type ProblemDaemonConfigMap map[ProblemDaemonType]*[]string
```
Done.

xueweiz commented on cmd/options/options.go:
```diff
@@ -53,14 +50,25 @@ type NodeProblemDetectorOptions struct {
+	SystemLogMonitorConfigPaths []string
+	// CustomPluginMonitorConfigPaths specifies the list of paths to custom plugin monitor configuration
+	// files.
+	CustomPluginMonitorConfigPaths []string
```
Done. Fixed in commit "Add existing monitors into the problem daemon registration hook".

xueweiz replied:
Sorry I added the explaination in the AddFlags(). I forgot to add it in the comments. Just added it. Thanks!

xueweiz commented on cmd/options/options.go:
```diff
@@ -79,6 +87,17 @@ func (npdo *NodeProblemDetectorOptions) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&npdo.ServerAddress, "address",
 		"127.0.0.1", "The address to bind the node problem detector server.")
+
+	for _, problemDaemonName := range problemdaemon.GetProblemDaemonNames() {
```
Done. Fixed in commit "Add existing monitors into the problem daemon registration hook".
Let's panic when both options are specified.

xueweiz commented on pkg/types/types.go:
```diff
@@ -113,3 +113,20 @@ type Exporter interface {
+// ProblemDaemonHandler represents the initialization handler for a type problem daemon.
+type ProblemDaemonHandler struct {
+	// Factory initializes a problem daemon instance.
+	Factory func(string) Monitor
```
Done.

xueweiz commented:
The Godep dependency on OpenCensus and gopsutil was broken, so this PR won't compile.
I uploaded #289  to fix the dependency.

Since #289 is not merged yet, I temporarily also include it in this PR, just so that travis can build.
When #289 is merged, I will then rebase to head, and remove the 70k+ changes from this PR.

xueweiz commented:
Finally I got some useful information on the TestGoroutineLeak flake:
See [pull-npd-test #1136427251437408257](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/node-problem-detector/275/pull-npd-test/1136427251437408257):

```
=== RUN   TestGoroutineLeak
--- FAIL: TestGoroutineLeak (0.01s)
    log_watcher_test.go:206: 
        	Error Trace:	log_watcher_test.go:206
        	Error:      	Not equal: 
        	            	expected: 9
        	            	actual  : 8
        	Test:       	TestGoroutineLeak
        	Messages:   	original stacktraces:
        	            	goroutine profile: total 9
        	            	3 @ 0x49f215 0x49c0c8 0x499ecb 0x602c03 0x487111
        	            	#	0x49f214	syscall.Syscall+0x4					/usr/local/go/src/syscall/asm_linux_amd64.s:18
        	            	#	0x49c0c7	syscall.read+0x67					/usr/local/go/src/syscall/zsyscall_linux_amd64.go:732
        	            	#	0x499eca	syscall.Read+0x5a					/usr/local/go/src/syscall/syscall_unix.go:172
        	            	#	0x602c02	github.com/sigma/go-inotify.(*Watcher).readEvents+0x512	/home/prow/go/src/k8s.io/node-problem-detector/vendor/github.com/sigma/go-inotify/inotify_linux.go:160
        	            	
        	            	1 @ 0x43c95d 0x60251b 0x60ede6 0x60e600 0x487111
        	            	#	0x60251a	github.com/sigma/go-inotify.(*Watcher).RemoveWatch+0x27a	/home/prow/go/src/k8s.io/node-problem-detector/vendor/github.com/sigma/go-inotify/inotify_linux.go:146
        	            	#	0x60ede5	github.com/google/cadvisor/utils/tail.(*Tail).watchFile+0x6c5	/home/prow/go/src/k8s.io/node-problem-detector/vendor/github.com/google/cadvisor/utils/tail/tail.go:159
        	            	#	0x60e5ff	github.com/google/cadvisor/utils/tail.(*Tail).watchLoop+0x3f	/home/prow/go/src/k8s.io/node-problem-detector/vendor/github.com/google/cadvisor/utils/tail/tail.go:128
        	            	
        	            	1 @ 0x458e7f 0x42dff3 0x42dfc9 0x42dd45 0x602b6f 0x487111
        	            	#	0x602b6e	github.com/sigma/go-inotify.(*Watcher).readEvents+0x47e	/home/prow/go/src/k8s.io/node-problem-detector/vendor/github.com/sigma/go-inotify/inotify_linux.go:215
        	            	
        	            	1 @ 0x458e7f 0x42ed19 0x42ecef 0x42ea8b 0x567973 0x56da99 0x567264 0x5698c4 0x56811c 0xe22233 0x458a6c 0x487111
        	            	#	0x567972	testing.(*T).Run+0x692		/usr/local/go/src/testing/testing.go:917
        	            	#	0x56da98	testing.runTests.func1+0xa8	/usr/local/go/src/testing/testing.go:1157
        	            	#	0x567263	testing.tRunner+0x163		/usr/local/go/src/testing/testing.go:865
        	            	#	0x5698c3	testing.runTests+0x523		/usr/local/go/src/testing/testing.go:1155
        	            	#	0x56811b	testing.(*M).Run+0x2eb		/usr/local/go/src/testing/testing.go:1072
        	            	#	0xe22232	main.main+0x222			_testmain.go:46
        	            	#	0x458a6b	runtime.main+0x20b		/usr/local/go/src/runtime/proc.go:200
        	            	
        	            	1 @ 0x458e7f 0x42ed19 0x42ecef 0x42eacb 0x5fd0ee 0x487111
        	            	#	0x5fd0ed	github.com/golang/glog.(*loggingT).flushDaemon+0xad	/home/prow/go/src/k8s.io/node-problem-detector/vendor/github.com/golang/glog/glog.go:882
        	            	
        	            	1 @ 0x458e7f 0x42ed19 0x42ecef 0x42eacb 0x608bce 0x487111
        	            	#	0x608bcd	k8s.io/klog.(*loggingT).flushDaemon+0xad	/home/prow/go/src/k8s.io/node-problem-detector/vendor/k8s.io/klog/klog.go:1035
        	            	
        	            	1 @ 0x5e5bee 0x5e59ba 0x5e186c 0xe1fa64 0x567264 0x487111
        	            	#	0x5e5bed	runtime/pprof.writeRuntimeProfile+0x9d								/usr/local/go/src/runtime/pprof/pprof.go:708
        	            	#	0x5e59b9	runtime/pprof.writeGoroutine+0xc9								/usr/local/go/src/runtime/pprof/pprof.go:670
        	            	#	0x5e186b	runtime/pprof.(*Profile).WriteTo+0x4fb								/usr/local/go/src/runtime/pprof/pprof.go:329
        	            	#	0xe1fa63	k8s.io/node-problem-detector/pkg/systemlogmonitor/logwatchers/filelog.TestGoroutineLeak+0xd3	/home/prow/go/src/k8s.io/node-problem-detector/pkg/systemlogmonitor/logwatchers/filelog/log_watcher_test.go:189
        	            	#	0x567263	testing.tRunner+0x163										/usr/local/go/src/testing/testing.go:865
        	            	
        	            	
        	            	end stacktraces:
        	            	goroutine profile: total 8
        	            	3 @ 0x49f215 0x49c0c8 0x499ecb 0x602c03 0x487111
        	            	#	0x49f214	syscall.Syscall+0x4					/usr/local/go/src/syscall/asm_linux_amd64.s:18
        	            	#	0x49c0c7	syscall.read+0x67					/usr/local/go/src/syscall/zsyscall_linux_amd64.go:732
        	            	#	0x499eca	syscall.Read+0x5a					/usr/local/go/src/syscall/syscall_unix.go:172
        	            	#	0x602c02	github.com/sigma/go-inotify.(*Watcher).readEvents+0x512	/home/prow/go/src/k8s.io/node-problem-detector/vendor/github.com/sigma/go-inotify/inotify_linux.go:160
        	            	
        	            	1 @ 0x458e7f 0x42dff3 0x42dfc9 0x42dd45 0x602b6f 0x487111
        	            	#	0x602b6e	github.com/sigma/go-inotify.(*Watcher).readEvents+0x47e	/home/prow/go/src/k8s.io/node-problem-detector/vendor/github.com/sigma/go-inotify/inotify_linux.go:215
        	            	
        	            	1 @ 0x458e7f 0x42ed19 0x42ecef 0x42ea8b 0x567973 0x56da99 0x567264 0x5698c4 0x56811c 0xe22233 0x458a6c 0x487111
        	            	#	0x567972	testing.(*T).Run+0x692		/usr/local/go/src/testing/testing.go:917
        	            	#	0x56da98	testing.runTests.func1+0xa8	/usr/local/go/src/testing/testing.go:1157
        	            	#	0x567263	testing.tRunner+0x163		/usr/local/go/src/testing/testing.go:865
        	            	#	0x5698c3	testing.runTests+0x523		/usr/local/go/src/testing/testing.go:1155
        	            	#	0x56811b	testing.(*M).Run+0x2eb		/usr/local/go/src/testing/testing.go:1072
        	            	#	0xe22232	main.main+0x222			_testmain.go:46
        	            	#	0x458a6b	runtime.main+0x20b		/usr/local/go/src/runtime/proc.go:200
        	            	
        	            	1 @ 0x458e7f 0x42ed19 0x42ecef 0x42eacb 0x5fd0ee 0x487111
        	            	#	0x5fd0ed	github.com/golang/glog.(*loggingT).flushDaemon+0xad	/home/prow/go/src/k8s.io/node-problem-detector/vendor/github.com/golang/glog/glog.go:882
        	            	
        	            	1 @ 0x458e7f 0x42ed19 0x42ecef 0x42eacb 0x608bce 0x487111
        	            	#	0x608bcd	k8s.io/klog.(*loggingT).flushDaemon+0xad	/home/prow/go/src/k8s.io/node-problem-detector/vendor/k8s.io/klog/klog.go:1035
        	            	
        	            	1 @ 0x5e5bee 0x5e59ba 0x5e186c 0xe1fd94 0x567264 0x487111
        	            	#	0x5e5bed	runtime/pprof.writeRuntimeProfile+0x9d								/usr/local/go/src/runtime/pprof/pprof.go:708
        	            	#	0x5e59b9	runtime/pprof.writeGoroutine+0xc9								/usr/local/go/src/runtime/pprof/pprof.go:670
        	            	#	0x5e186b	runtime/pprof.(*Profile).WriteTo+0x4fb								/usr/local/go/src/runtime/pprof/pprof.go:329
        	            	#	0xe1fd93	k8s.io/node-problem-detector/pkg/systemlogmonitor/logwatchers/filelog.TestGoroutineLeak+0x403	/home/prow/go/src/k8s.io/node-problem-detector/pkg/systemlogmonitor/logwatchers/filelog/log_watcher_test.go:204
        	            	#	0x567263	testing.tRunner+0x163										/usr/local/go/src/testing/testing.go:865
```

Most importantly, there is a goroutine that exists at the beginning, but disappears at the end, which caused the test to fail:
```
        	            	1 @ 0x43c95d 0x60251b 0x60ede6 0x60e600 0x487111
        	            	#	0x60251a	github.com/sigma/go-inotify.(*Watcher).RemoveWatch+0x27a	/home/prow/go/src/k8s.io/node-problem-detector/vendor/github.com/sigma/go-inotify/inotify_linux.go:146
        	            	#	0x60ede5	github.com/google/cadvisor/utils/tail.(*Tail).watchFile+0x6c5	/home/prow/go/src/k8s.io/node-problem-detector/vendor/github.com/google/cadvisor/utils/tail/tail.go:159
        	            	#	0x60e5ff	github.com/google/cadvisor/utils/tail.(*Tail).watchLoop+0x3f	/home/prow/go/src/k8s.io/node-problem-detector/vendor/github.com/google/cadvisor/utils/tail/tail.go:128
```

@Random-Liu @andyxning @wangzhen127 Have you seen similar failures before? I don't yet know how this is related to my change. I suspect this or similar goroutine fluctuations could cause flakes all along.
I think maybe we could remove the TestGoroutineLeak test, and add e2e stress/soak tests to NPD. Because I don't think a unit test is best suited for testing leakage anyways. WDYT?

xueweiz commented on pkg/problemdaemon/problem_daemon.go:
```diff
@@ -0,0 +1,58 @@
+			problemDaemon := handlers[problemDaemonType].Factory(config)
+			if problemDaemon != nil {
+				problemDaemons = append(problemDaemons, problemDaemon)
+			}
```
Good catch. Now the expectation is any error should result in panic. So this code path is useless, and is removed.

xueweiz commented:
/assign @wangzhen127

wangzhen127 commented on pkg/systemlogmonitor/log_monitor_test.go:
```diff
@@ -151,26 +145,3 @@ func TestGenerateStatus(t *testing.T) {
 	}
 }
-
-func TestGoroutineLeak(t *testing.T) {
```
Remember to drop this commit.

xueweiz replied:
Will keep that in mind.

wangzhen127 replied:
I think you still need to drop this commit, right? :)

xueweiz replied:
Done.

wangzhen127 commented on cmd/node_problem_detector.go:
```diff
@@ -69,32 +50,41 @@ func main() {
+	// Legacy per-monitor configuration.
+	// This section is kept so that existing command line option still works on NPD.
+	// TODO(xueweiz): when NPD's refactoring is finished, remove legacy configuration.
 	for _, config := range npdo.SystemLogMonitorConfigPaths {
```
s/config/configPath

xueweiz replied:
Done.

wangzhen127 commented on cmd/node_problem_detector.go:
```diff
@@ -69,32 +50,41 @@ func main() {
+	// This section is kept so that existing command line option still works on NPD.
+	// TODO(xueweiz): when NPD's refactoring is finished, remove legacy configuration.
 	for _, config := range npdo.SystemLogMonitorConfigPaths {
-		if _, ok := monitors[config]; ok {
```
I think we still need to keep this check? For the case: `-- system-log-monitor=config1,config1,config2`. Notice `config1` is specified twice.

xueweiz replied:
Good point. I just merged the two sections of initialization code together. `problemdaemon.NewProblemDaemons()` is handling this logic now.

wangzhen127 commented on pkg/problemdaemon/problem_daemon.go:
```diff
@@ -0,0 +1,55 @@
+	problemDaemons := []types.Monitor{}
+	for problemDaemonType, configs := range monitorConfigPaths {
+		for _, config := range *configs {
+			problemDaemons = append(problemDaemons, handlers[problemDaemonType].CreateProblemDaemonOrDie(config))
```
We need to check if there are duplicated configs. For the case: `--config.system-log-monitor=config1,config1,config2`. Notice that `config1` is specified twice here. The duplicated one should be ignored. Also see the other similar comment for legacy flags in `cmd/node_problem_detector.go`.

xueweiz replied:
Good point. Done.

wangzhen127 commented on pkg/exporters/k8sexporter/k8s_exporter.go:
```diff
@@ -0,0 +1,92 @@
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte("ok"))
+	})
+
```
Add `// Add the handler to serve condition http request.` (Just to keep the comment from the original code)

xueweiz replied:
Good catch. Done.

xueweiz commented on cmd/options/options.go:
```diff
@@ -47,41 +42,87 @@ type NodeProblemDetectorOptions struct {
-	if len(npdo.SystemLogMonitorConfigPaths) == 0 && len(npdo.CustomPluginMonitorConfigPaths) == 0 {
-		panic(fmt.Sprintf("Either --system-log-monitors or --custom-plugin-monitors is required"))
+
+	if len(npdo.SystemLogMonitorConfigPaths) != 0 && npdo.MonitorConfigPaths["system-log-monitor"] != nil && len(*npdo.MonitorConfigPaths["system-log-monitor"]) != 0 {
```
Done.

xueweiz commented on cmd/options/options.go:
```diff
@@ -47,41 +42,87 @@ type NodeProblemDetectorOptions struct {
+	if len(npdo.SystemLogMonitorConfigPaths) != 0 && npdo.MonitorConfigPaths["system-log-monitor"] != nil && len(*npdo.MonitorConfigPaths["system-log-monitor"]) != 0 {
+		panic("Option --system-log-monitors is deprecated in favor of --config.system-log-monitor. They cannot be set at the same time.")
+	}
+	if len(npdo.CustomPluginMonitorConfigPaths) != 0 && npdo.MonitorConfigPaths["system-log-monitor"] != nil && len(*npdo.MonitorConfigPaths["custom-plugin-monitor"]) != 0 {
```
Done.

xueweiz commented on cmd/options/options_test.go:
```diff
@@ -82,3 +86,134 @@ func TestSetNodeNameOrDie(t *testing.T) {
+	}
+}
+
+// // ValidOrDie validates node problem detector command line options.
```
Done. Sorry.

xueweiz commented on cmd/node_problem_detector.go:
```diff
@@ -69,32 +50,41 @@ func main() {
-	if npdo.ServerPort > 0 {
-		startHTTPServer(p, npdo)
+	if len(problemDaemons) == 0 {
+		glog.Fatalf("No problem daemon is successfully setup")
```
Done.

xueweiz commented:
@wangzhen127 All above comments are fixed. Could you help take a look again? Thanks!

wangzhen127 commented on cmd/node_problem_detector.go:
```diff
@@ -66,35 +45,35 @@ func main() {
-		monitors[config] = systemlogmonitor.NewLogMonitorOrDie(config)
+	// Initialize problem daemons.
+	problemDaemons := problemdaemon.NewProblemDaemons(npdo.MonitorConfigPaths)
+
```
Nit: Remove blank line.

xueweiz replied:
Done.

wangzhen127 commented on cmd/node_problem_detector.go:
```diff
@@ -66,35 +45,35 @@ func main() {
 	}
-	c := problemclient.NewClientOrDie(npdo)
-	p := problemdetector.NewProblemDetector(monitors, c)
 
```
Nit: Remove blank line.

xueweiz replied:
Done.

wangzhen127 commented on pkg/systemstatsmonitor/system_stats_monitor.go:
```diff
@@ -0,0 +1,110 @@
+func init() {
+	problemdaemon.Register("system-stats-monitor", types.ProblemDaemonHandler{
+		CreateProblemDaemonOrDie: NewSystemStatsMonitorOrDie,
+		CmdOptionDescription:     "Set to true to enable, or any other value to disable."})
```
Should this be: `Set to config file paths`?

xueweiz replied:
Good catch, thanks!
Done.

wangzhen127 commented on pkg/problemdaemon/problem_daemon.go:
```diff
@@ -0,0 +1,67 @@
+
+// GetProblemDaemonHandler retrieves the ProblemDaemonHandler for a specific type of problem daemon.
+func GetProblemDaemonHandler(problemDaemonType types.ProblemDaemonType) types.ProblemDaemonHandler {
+	return handlers[problemDaemonType]
```
What happens when `problemDaemonType` is not in the map? This probably never happen for now. But later if we can disable some monitors, this needs to be checked. It is better to handle that now. Maybe panic?

xueweiz replied:
Good point. Done.

wangzhen127 commented on pkg/problemdaemon/problem_daemon.go:
```diff
@@ -0,0 +1,67 @@
+
+// GetProblemDaemonNames retrieves the names of all available problem daemon types.
+func GetProblemDaemonNames() []types.ProblemDaemonType {
+	problemDaemons := []types.ProblemDaemonType{}
```
Nit: s/problemDaemons/problemDaemonTypes

xueweiz replied:
Done.

wangzhen127 commented on pkg/custompluginmonitor/custom_plugin_monitor_test.go:
```diff
@@ -0,0 +1,29 @@
+)
+
+func TestRegistration(t *testing.T) {
+	if problemdaemon.GetProblemDaemonHandler("custom-plugin-monitor").CmdOptionDescription == "" {
```
Checking the cmd option desc feels weird. What if some daemon does not have any customized description? Should we check `problemdaemon.GetProblemDaemonHandler("custom-plugin-monitor")` returns non nil or panic or something like that? This depends on the behavior of `GetProblemDaemonHandler()`.

Same question applies to other daemon registration tests.

xueweiz replied:
Done.

wangzhen127 commented on pkg/systemstatsmonitor/types/config_test.go:
```diff
@@ -0,0 +1,142 @@
+	"time"
+)
+
+func TestApplyConfiguration(t *testing.T) {
```
Try to use the table-driven tests with subtests for all the tests you are adding.

xueweiz replied:
Done.

xueweiz commented on cmd/options/options.go:
```diff
@@ -47,41 +44,132 @@ type NodeProblemDetectorOptions struct {
+}
+
+// SetConfigFromDeprecatedOptionsOrDie sets NPD option using deprecated options.
+func (npdo *NodeProblemDetectorOptions) SetConfigFromDeprecatedOptionsOrDie() {
```
Done.

xueweiz commented on cmd/options/options.go:
```diff
@@ -47,41 +44,132 @@ type NodeProblemDetectorOptions struct {
-		panic(fmt.Sprintf("Either --system-log-monitors or --custom-plugin-monitors is required"))
+
+	if len(npdo.SystemLogMonitorConfigPaths) != 0 {
+		panic("SystemLogMonitorConfigPaths is deprecated. This should not happen. It must be a NPD bug.")
```
Done.

xueweiz commented on cmd/options/options.go:
```diff
@@ -47,41 +44,132 @@ type NodeProblemDetectorOptions struct {
+		panic("SystemLogMonitorConfigPaths is deprecated. This should not happen. It must be a NPD bug.")
+	}
+	if len(npdo.CustomPluginMonitorConfigPaths) != 0 {
+		panic("CustomPluginMonitorConfigPaths is deprecated. This should not happen. It must be a NPD bug.")
```
Done.

xueweiz commented on cmd/node_problem_detector.go:
```diff
@@ -66,35 +45,35 @@ func main() {
 
 	npdo.SetNodeNameOrDie()
+	npdo.SetConfigFromDeprecatedOptionsOrDie()
 
```
Don.e

xueweiz commented on cmd/options/options_test.go:
```diff
@@ -82,3 +86,98 @@ func TestSetNodeNameOrDie(t *testing.T) {
+	emptyMonitorConfigMap := types.ProblemDaemonConfigPathMap{}
+
+	for _, test := range []struct {
+		npdo        NodeProblemDetectorOptions
```
Done.

xueweiz commented on cmd/options/options_test.go:
```diff
@@ -82,3 +86,98 @@ func TestSetNodeNameOrDie(t *testing.T) {
+			npdo: NodeProblemDetectorOptions{
+				SystemLogMonitorConfigPaths: []string{"config-a"},
+			},
+			expectPanic: true,
```
Whenever deprecated fields got used (i.e. `SystemLogMonitorConfigPaths`, `CustomPluginMonitorConfigPaths`), it will be considered as invalid, and `ValidOrDie()` will panic.

xueweiz replied:
Done.

xueweiz commented:
@wangzhen127 @Random-Liu Thanks for the previous review! I just made the changes. Could you help take a look again? Thanks!

wangzhen127 commented on cmd/options/options_test.go:
```diff
@@ -18,9 +18,44 @@ package options
+func equalMonitorConfigPaths(npdoX NodeProblemDetectorOptions, npdoY NodeProblemDetectorOptions) bool {
+	monitorConfigPathsX, monitorConfigPathsY := npdoX.MonitorConfigPaths, npdoY.MonitorConfigPaths
+
+	if monitorConfigPathsX != nil || monitorConfigPathsY != nil {
```
Nit: There are a lot of indentation in the nested `if` and `for` clauses. Usually, a better style is to return early. For example:
```
if monitorConfigPathsX == nil && monitorConfigPathsY == nil {
    return true
}
if monitorConfigPathsX == nil || monitorConfigPathsY == nil {
    return false
}
if len(monitorConfigPathsX) != len(monitorConfigPathsY) {
    return false
}
for ...
```

xueweiz replied:
Sorry for the rookie mistake. I was originally designing this function to compare the equality between npdoX and npdoY, hence wasn't returning early. Later on when I changed the intention, I forgot to change the code.
Done.

wangzhen127 commented on cmd/options/options_test.go:
```diff
@@ -18,9 +18,44 @@ package options
+			if !ok {
+				return false
+			}
+			if configPathsX != nil || configPathsY != nil {
```
Similarly, here you can do:
```
if configPathsX == nil && configPathsY == nil {
    continue
}
if configPathsX == nil || configPathsY == nil {
    return false
}
if !reflect.DeepEqual(*configPathsX, *configPathsY) {
    return false
}
```

xueweiz replied:
Sorry. Done.

wangzhen127 commented on pkg/systemlogmonitor/log_monitor_test.go:
```diff
@@ -145,11 +153,24 @@ func TestGenerateStatus(t *testing.T) {
 
 func TestGoroutineLeak(t *testing.T) {
-	orignal := runtime.NumGoroutine()
+	original_prof := pprof.Lookup("goroutine")
```
Should those changes to `TestGoroutineLeak` be moved to the other PR?

xueweiz replied:
Done.

wangzhen127 commented on pkg/systemlogmonitor/logwatchers/filelog/log_watcher_test.go:
```diff
@@ -182,7 +183,12 @@ Jan  2 03:04:05 kernel: [2.000000] 3
 
 func TestGoroutineLeak(t *testing.T) {
-	orignal := runtime.NumGoroutine()
+	original_prof := pprof.Lookup("goroutine")
```
Should those changes to `TestGoroutineLeak` be moved to the other PR?

xueweiz replied:
Done.

xueweiz commented on README.md:
```diff
@@ -59,6 +59,12 @@ List of supported problem daemons:
 | [KernelMonitor](https://github.com/kubernetes/node-problem-detector/blob/master/config/kernel-monitor.json) | KernelDeadlock | A system log monitor monitors kernel log and reports problem according to predefined rules. |
 | [AbrtAdaptor](https://github.com/kubernetes/node-problem-detector/blob/master/config/abrt-adaptor.json) | None | Monitor ABRT log messages and report them further. ABRT (Automatic Bug Report Tool) is health monitoring daemon able to catch kernel problems as well as application crashes of various kinds occurred on the host. For more information visit the [link](https://github.com/abrt). |
 | [CustomPluginMonitor](https://github.com/kubernetes/node-problem-detector/blob/master/config/custom-plugin-monitor.json) | On-demand(According to users configuration) | A custom plugin monitor for node-problem-detector to invoke and check various node problems with user defined check scripts. See proposal [here](https://docs.google.com/document/d/1jK_5YloSYtboj-DtfjmYKxfNnUxCAvohLnsH5aGCAYQ/edi ...
+| [SystemStatsMonitor](https://github.com/kubernetes/node-problem-detector/blob/master/config/system-stats-monitor.json) | None(Could be added in future) | A system stats monitor for node-problem-detector to collect various health-related system stats. See proposal [here](https://docs.google.com/document/d/1SeaUz6kBavI283Dq8GBpoEUDrHA2a795xtw0OvjM568/edit). |
```
Thanks! Done.

xueweiz commented on README.md:
```diff
@@ -59,6 +59,12 @@ List of supported problem daemons:
+
+# Exporter
+
+An exporter is a component of node-problem-detector. It report node problems and/or metrics to 
```
Sorry. Done.

xueweiz commented on README.md:
```diff
@@ -59,6 +59,12 @@ List of supported problem daemons:
+# Exporter
+
+An exporter is a component of node-problem-detector. It report node problems and/or metrics to 
+certain control plane (e.g. Kubernetes API server, or Prometheus scrape endpoint).
```
Good point. Done.

xueweiz commented on README.md:
````diff
@@ -149,12 +168,13 @@ For example, to test [KernelMonitor](https://github.com/kubernetes/node-problem-
 2. ```kubectl proxy --port=8080``` (make a running cluster's API server available locally)
 3. Update [KernelMonitor](https://github.com/kubernetes/node-problem-detector/blob/master/config/kernel-monitor.json)'s ```logPath``` to your local kernel log directory. For example, on some Linux systems, it is ```/run/log/journal``` instead of ```/var/log/journal```.
-3. ```./bin/node-problem-detector --logtostderr --apiserver-override=http://127.0.0.1:8080?inClusterConfig=false --system-log-monitors=config/kernel-monitor.json  --port=20256``` (or point to any API server address:port)
+3. ```./bin/node-problem-detector --logtostderr --apiserver-override=http://127.0.0.1:8080?inClusterConfig=false --config.system-log-monitor=config/kernel-monitor.json --config.system-stats-monitor=config/system-stats-monitor.json --port=20256 --prometheus-port=20257``` (or point to any API server address:port)
````
Done.

xueweiz commented on pkg/systemlogmonitor/README.md:
```diff
@@ -1,7 +1,25 @@
-# System Log Monitor
```
...Sorry...Must be a manual error during copy & paste. Done.

xueweiz commented on cmd/options/options_test.go:
```diff
@@ -82,3 +117,202 @@ func TestSetNodeNameOrDie(t *testing.T) {
+			},
+		},
+		{
+			name: "using deprecated & new options on different problem daemons",
```
Done.

xueweiz replied:
Done.

xueweiz commented on cmd/options/options_test.go:
```diff
@@ -82,3 +117,202 @@ func TestSetNodeNameOrDie(t *testing.T) {
+			},
+		},
+		{
+			name: "using deprecated & new options on same problem daemon",
```
Done.

xueweiz commented:
@wangzhen127 Thanks for the review Zhen! Just address them. Could you help take a look again? Thanks :)

Random-Liu commented:
/test pull-npd-test

Random-Liu commented on cmd/options/options.go:
```diff
@@ -47,41 +44,132 @@ type NodeProblemDetectorOptions struct {
 
 // AddFlags adds node problem detector command line options to pflag.
 func (npdo *NodeProblemDetectorOptions) AddFlags(fs *pflag.FlagSet) {
 	fs.StringSliceVar(&npdo.SystemLogMonitorConfigPaths, "system-log-monitors",
```
MarkDeprecated?

xueweiz replied:
Sorry, what do you mean by "MarkDeprecated?" i.e. add a comment? Or is there some good procedure we could follow?
Currently I'm documenting that it's deprecated in the help text: `This option is deprecated, replaced by --config.system-log-monitor, and will be removed. NPD will panic if both --system-log-monitors and --config.system-log-monitor are set.`

xueweiz replied:
Changed to use MarkDeprecated to handle this logic.
Now, when user use the deprecated option, a warning will be printed as this:

> Flag --system-log-monitors has been deprecated, replaced by --config.system-log-monitor. NPD will panic if both --system-log-monitors and --config.system-log-monitor are set.

Random-Liu commented on cmd/options/options.go:
```diff
@@ -47,41 +44,132 @@ type NodeProblemDetectorOptions struct {
 	fs.StringSliceVar(&npdo.SystemLogMonitorConfigPaths, "system-log-monitors",
-		[]string{}, "List of paths to system log monitor config files, comma separated.")
+		[]string{}, "List of paths to system log monitor config files, comma separated. This option is deprecated, replaced by --config.system-log-monitor, and will be removed. NPD will panic if both --system-log-monitors and --config.system-log-monitor are set.")
 	fs.StringSliceVar(&npdo.CustomPluginMonitorConfigPaths, "custom-plugin-monitors",
```
MarkDeprecated?

xueweiz replied:
Done.

Random-Liu commented on cmd/node_problem_detector.go:
```diff
@@ -17,42 +17,21 @@ limitations under the License.
-		}
-	}()
-}
+	_ "k8s.io/node-problem-detector/pkg/systemstatsmonitor"
```
Can we create a plugins.go in the main package, which just import plugins?

Like this https://github.com/containerd/containerd/blob/master/cmd/containerd/builtins_linux.go

xueweiz replied:
Sounds good :) Done.

Random-Liu commented on cmd/options/options.go:
```diff
@@ -24,20 +24,17 @@ import (
 
 	"github.com/spf13/pflag"
+
+	"k8s.io/node-problem-detector/pkg/custompluginmonitor"
```
This makes it impossible to compile out custompluginmonitor and systemlogmonitor.

I think to handle the deprecated flag, it is fine to hard code the plugin name in this file to get rid of this unnecessary dependency.

xueweiz replied:
After this PR is merged, I plan to change the Prow tests that's still using old NPD flags.

Once the testing changes are done, I plan to remove the deprecated flags and all the logic around it (including here). And then we will have the ability to compile out custompluginmonitor and systemlogmonitor.

By "remove", I mean that when you do `./node_problem_detector --help`, those flags won't show up at all. It's like they never existed.

Random-Liu replied:
You can use `MarkDeprecated` to achieve that, and we should keep those flags for backward compatibility. See https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli

xueweiz replied:
Sounds good. Thanks!

Random-Liu replied:
We should really remove this. :p 

It is unnecessary to break the compile-in plugin system just because of a constant.

I'll send a PR to address this. :)

xueweiz commented:
Hi Lantao, thanks for the review! Just addressed the comments and rebased on top of #290 . However I didn't quite understand the "MarkDeprecated" part. Could you help elaborate a little bit? Thanks!

wangzhen127 commented:
I think he means https://godoc.org/github.com/spf13/pflag#FlagSet.MarkDeprecated

xueweiz commented:
Thanks for the tips on the MarkDeprecated function! Just made the adjustments.
@wangzhen127  @Random-Liu , could you help take a look again? Thanks!

wangzhen127 commented:
/lgtm
/hold
Wait for @Random-Liu's approval.

Random-Liu commented:
/hold cancel
/lgtm
